File: packages/rayact-webview/android/src/main/java/dev/rayact/webview/RayactWebViewRegistration.kt
package dev.rayact.webview

import android.content.Context
import android.graphics.Color
import android.webkit.JavascriptInterface
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.webkit.ScriptHandler
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import com.rayact.engine.RayactPlatformModuleRegistration
import com.rayact.engine.RayactPlatformRegistry
import com.rayact.engine.RayactPlatformViewContext
import com.rayact.engine.RayactPlatformViewController
import org.json.JSONObject

class RayactWebViewRegistration : RayactPlatformModuleRegistration {
    override fun register(context: Context, registry: RayactPlatformRegistry) {
        registry.registerViewFactory("webview") { viewContext ->
            RayactWebViewController(viewContext)
        }
    }
}

private class RayactWebViewController(
    private val host: RayactPlatformViewContext,
) : RayactPlatformViewController {
    override val view = WebView(host.context)
    private val properties = mutableMapOf<String, Any?>()
    private var disposed = false
    private var documentStartHandler: ScriptHandler? = null
    private var documentStartFallback: String? = null

    init {
        // Android WebView otherwise paints its own opaque default canvas even
        // when the loaded document has no CSS background. Keep the native
        // surface clear so Rayact content underneath remains visible; an
        // explicit background supplied by the document still paints normally.
        view.setBackgroundColor(Color.TRANSPARENT)
        view.background?.alpha = 0
        view.settings.javaScriptEnabled = true
        view.settings.blockNetworkLoads = false
        view.addJavascriptInterface(object {
            @JavascriptInterface
            fun postMessage(message: String) {
                emit("message", message)
            }
        }, "RayactWebView")
        view.webViewClient = object : WebViewClient() {
            override fun onPageStarted(
                view: WebView?,
                url: String?,
                favicon: android.graphics.Bitmap?,
            ) {
                documentStartFallback?.let { view?.evaluateJavascript(it, null) }
                emit("loadStart", url.orEmpty())
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                (properties["injectedJavaScript"] as? String)
                    ?.takeIf(String::isNotEmpty)
                    ?.let { view?.evaluateJavascript(it, null) }
                emit("loadEnd", url.orEmpty())
                emit(
                    "navigationStateChange",
                    JSONObject()
                        .put("url", url.orEmpty())
                        .put("canGoBack", view?.canGoBack() == true)
                        .put("canGoForward", view?.canGoForward() == true)
                        .toString(),
                )
            }

            override fun onReceivedError(
                view: WebView?,
                request: WebResourceRequest?,
                error: WebResourceError?,
            ) {
                if (request?.isForMainFrame != false) {
                    emit("error", error?.description?.toString().orEmpty())
                }
            }
        }
        applyProperties(host.initialProperties)
    }

    private fun emit(type: String, data: String = "") {
        if (disposed) return
        host.emit(JSONObject().put("type", type).put("data", data).toString())
    }

    private fun loadSource() {
        if (disposed) return
        (properties["sourceHtml"] as? String)?.let {
            view.loadDataWithBaseURL(
                properties["baseUrl"] as? String,
                it,
                "text/html",
                "UTF-8",
                null,
            )
            return
        }
        (properties["sourceUri"] as? String)?.takeIf(String::isNotEmpty)?.let(view::loadUrl)
    }

    override fun setProperties(properties: Map<String, Any?>) {
        applyProperties(properties)
    }

    private fun applyProperties(changes: Map<String, Any?>) {
        if (disposed) return
        for ((key, value) in changes) {
            if (value == null) properties.remove(key) else properties[key] = value
        }
        // Source variants are mutually exclusive even when a producer omits the
        // null tombstone. This prevents a previous HTML value from shadowing a
        // later URI (and vice versa).
        if (changes["sourceHtml"] is String) {
            properties.remove("sourceUri")
        } else if (changes["sourceUri"] is String) {
            properties.remove("sourceHtml")
            properties.remove("baseUrl")
        }

        // Configuration is committed before a changed source begins loading.
        if ("javaScriptEnabled" in changes) {
            view.settings.javaScriptEnabled = booleanProperty("javaScriptEnabled", true)
        }
        if ("scrollEnabled" in changes) {
            val enabled = booleanProperty("scrollEnabled", true)
            view.isVerticalScrollBarEnabled = enabled
            view.isHorizontalScrollBarEnabled = enabled
        }
        if ("injectedJavaScriptBeforeContentLoaded" in changes) {
            rebuildDocumentStartScript()
        }
        if (changes.keys.any { it == "sourceUri" || it == "sourceHtml" || it == "baseUrl" }) {
            view.post(::loadSource)
        }

        // Commands are deliberately applied after configuration and source.
        (changes["command"] as? String)?.takeIf(String::isNotEmpty)?.let(::runCommand)
    }

    private fun booleanProperty(key: String, fallback: Boolean): Boolean {
        return when (val value = properties[key]) {
            is Boolean -> value
            is Number -> value.toInt() != 0
            is String -> value != "false" && value != "0"
            else -> fallback
        }
    }

    private fun rebuildDocumentStartScript() {
        documentStartHandler?.remove()
        documentStartHandler = null
        documentStartFallback = null
        val script = (properties["injectedJavaScriptBeforeContentLoaded"] as? String)
            ?.takeIf(String::isNotEmpty)
            ?: return
        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            documentStartHandler =
                WebViewCompat.addDocumentStartJavaScript(view, script, setOf("*"))
        } else {
            // Older WebView providers have no document-start hook. Preserve
            // the previous best-effort behavior instead of dropping the prop.
            documentStartFallback = script
        }
    }

    private fun runCommand(raw: String) {
        val value = raw.replace(Regex("#\\d+$"), "")
        when {
            value == "reload" -> view.reload()
            value == "goBack" -> view.goBack()
            value == "goForward" -> view.goForward()
            value == "stopLoading" -> view.stopLoading()
            value.startsWith("inject:") ->
                view.evaluateJavascript(value.removePrefix("inject:"), null)
            value.startsWith("post:") -> {
                val message = JSONObject.quote(value.removePrefix("post:"))
                view.evaluateJavascript(
                    "window.dispatchEvent(new MessageEvent('message',{data:$message}));",
                    null,
                )
            }
        }
    }

    override fun dispose() {
        if (disposed) return
        disposed = true
        documentStartHandler?.remove()
        documentStartHandler = null
        documentStartFallback = null
        view.removeJavascriptInterface("RayactWebView")
        view.stopLoading()
        view.destroy()
    }
}

File: packages/rayact-webview/ios/RayactWebViewRegistration.swift
import Foundation
import WebKit

final class RayactWebViewRegistration: RayactPlatformModuleRegistration {
    func register(with registry: RayactPlatformRegistry) {
        registry.registerViewFactory("webview") { context in
            RayactWebViewController(context: context)
        }
    }
}

private final class RayactWebViewController: NSObject, RayactPlatformViewController, WKNavigationDelegate, WKScriptMessageHandler {
    let view: UIView

    private let context: RayactPlatformViewContext
    private let webView: WKWebView
    private var properties: [String: Any] = [:]
    private var disposed = false

    init(context: RayactPlatformViewContext) {
        self.context = context
        let configuration = WKWebViewConfiguration()
        configuration.userContentController.addUserScript(WKUserScript(
            source: """
            window.RayactWebView = window.RayactWebView || {};
            window.RayactWebView.postMessage = function(message) {
              window.webkit.messageHandlers.RayactWebView.postMessage(String(message));
            };
            """,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: false
        ))
        webView = WKWebView(frame: .zero, configuration: configuration)
        view = webView
        super.init()

        webView.isOpaque = false
        webView.backgroundColor = .clear
        webView.scrollView.backgroundColor = .clear
        if #available(iOS 15.0, *) {
            webView.underPageBackgroundColor = .clear
        }
        webView.navigationDelegate = self
        configuration.userContentController.add(self, name: "RayactWebView")
        applyProperties(context.initialProperties.mapValues(Optional.some))
    }

    func setProperties(_ properties: [String: Any?]) {
        applyProperties(properties)
    }

    private func applyProperties(_ changes: [String: Any?]) {
        guard !disposed else { return }
        for (key, value) in changes {
            if let value {
                properties[key] = value
            } else {
                properties.removeValue(forKey: key)
            }
        }
        // Keep the two source variants exclusive even if a producer does not
        // include the expected null tombstone in its patch.
        if changes["sourceHtml"] is String {
            properties.removeValue(forKey: "sourceUri")
        } else if changes["sourceUri"] is String {
            properties.removeValue(forKey: "sourceHtml")
            properties.removeValue(forKey: "baseUrl")
        }

        if changes.keys.contains("javaScriptEnabled") {
            webView.configuration.defaultWebpagePreferences.allowsContentJavaScript =
                boolProperty("javaScriptEnabled", fallback: true)
        }
        if changes.keys.contains("scrollEnabled") {
            webView.scrollView.isScrollEnabled = boolProperty("scrollEnabled", fallback: true)
        }
        if changes.keys.contains("injectedJavaScriptBeforeContentLoaded") {
            rebuildDocumentStartScripts()
        }
        if changes.keys.contains(where: {
            $0 == "sourceUri" || $0 == "sourceHtml" || $0 == "baseUrl"
        }) {
            loadSource()
        }
        if let command = changes["command"] as? String, !command.isEmpty {
            runCommand(command)
        }
    }

    private func rebuildDocumentStartScripts() {
        let controller = webView.configuration.userContentController
        controller.removeAllUserScripts()
        controller.addUserScript(WKUserScript(
            source: """
            window.RayactWebView = window.RayactWebView || {};
            window.RayactWebView.postMessage = function(message) {
              window.webkit.messageHandlers.RayactWebView.postMessage(String(message));
            };
            """,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: false
        ))
        if let script = properties["injectedJavaScriptBeforeContentLoaded"] as? String,
           !script.isEmpty {
            controller.addUserScript(WKUserScript(
                source: script,
                injectionTime: .atDocumentStart,
                forMainFrameOnly: true
            ))
        }
    }

    private func loadSource() {
        if let html = properties["sourceHtml"] as? String {
            let baseURL = (properties["baseUrl"] as? String).flatMap(URL.init(string:))
            webView.loadHTMLString(html, baseURL: baseURL)
            return
        }
        if let raw = properties["sourceUri"] as? String, let url = URL(string: raw) {
            webView.load(URLRequest(url: url))
        } else {
            webView.loadHTMLString("", baseURL: nil)
        }
    }

    private func boolProperty(_ key: String, fallback: Bool) -> Bool {
        switch properties[key] {
        case let value as Bool:
            return value
        case let value as NSNumber:
            return value.boolValue
        case let value as String:
            return value != "false" && value != "0"
        default:
            return fallback
        }
    }

    private func runCommand(_ raw: String) {
        let value = raw.replacingOccurrences(
            of: #"#[0-9]+$"#,
            with: "",
            options: .regularExpression
        )
        switch value {
        case "reload":
            // WKWebView.reload() re-requests the current URL. Inline HTML
            // loaded with loadHTMLString has no real URL (about:blank unless a
            // baseURL was given), so reload() would blank the page — Android's
            // WebView re-renders its data URL instead. Re-run the source load
            // for inline HTML so reload means the same thing on both platforms.
            if properties["sourceHtml"] is String {
                loadSource()
            } else {
                webView.reload()
            }
        case "goBack":
            webView.goBack()
        case "goForward":
            webView.goForward()
        case "stopLoading":
            webView.stopLoading()
        default:
            if value.hasPrefix("inject:") {
                webView.evaluateJavaScript(String(value.dropFirst("inject:".count)))
            } else if value.hasPrefix("post:") {
                let message = String(value.dropFirst("post:".count))
                let quoted = (try? JSONSerialization.data(withJSONObject: [message]))
                    .flatMap { String(data: $0, encoding: .utf8) }
                    .map { String($0.dropFirst().dropLast()) } ?? "\"\""
                webView.evaluateJavaScript(
                    "window.dispatchEvent(new MessageEvent('message',{data:\(quoted)}));"
                )
            }
        }
    }

    private func emit(_ type: String, data: String = "") {
        guard !disposed,
              let encoded = try? JSONSerialization.data(withJSONObject: [
                  "type": type,
                  "data": data,
              ]),
              let string = String(data: encoded, encoding: .utf8) else {
            return
        }
        context.emit(string)
    }

    func dispose() {
        guard !disposed else { return }
        disposed = true
        webView.stopLoading()
        webView.navigationDelegate = nil
        webView.configuration.userContentController.removeScriptMessageHandler(
            forName: "RayactWebView"
        )
        webView.removeFromSuperview()
    }

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        emit("message", data: String(describing: message.body))
    }

    func webView(
        _ webView: WKWebView,
        didStartProvisionalNavigation navigation: WKNavigation?
    ) {
        emit("loadStart", data: webView.url?.absoluteString ?? "")
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation?) {
        if let script = properties["injectedJavaScript"] as? String, !script.isEmpty {
            webView.evaluateJavaScript(script)
        }
        emit("loadEnd", data: webView.url?.absoluteString ?? "")
        let state: [String: Any] = [
            "url": webView.url?.absoluteString ?? "",
            "canGoBack": webView.canGoBack,
            "canGoForward": webView.canGoForward,
        ]
        if let data = try? JSONSerialization.data(withJSONObject: state),
           let string = String(data: data, encoding: .utf8) {
            emit("navigationStateChange", data: string)
        }
    }

    func webView(
        _ webView: WKWebView,
        didFailProvisionalNavigation navigation: WKNavigation?,
        withError error: Error
    ) {
        emit("error", data: error.localizedDescription)
    }

    func webView(
        _ webView: WKWebView,
        didFail navigation: WKNavigation?,
        withError error: Error
    ) {
        emit("error", data: error.localizedDescription)
    }
}
